File: Python/WebsiteCode/LeetCode/Medium/String/MakeStringASubsequenceUsingCyclicIncrements.py
# 13ms 17.57MB
class Solution:
    def canMakeSubsequence(self, str1: str, str2: str) -> bool:
        if len(str1)==len(str2):
            for i in range(len(str1)):
                if str2[i]=='a' and str1[i]=='z':
                    continue
                elif str1[i]==str2[i]:
                    continue
                elif ord(str2[i])-ord(str1[i])!=1:
                    return False
            return True

        l=[]
        if len(str2)>len(str1):
            return False
        for i in str2:
            if i not in str1:
                l.append(i)
        c=0
        for i in l:
            if chr(ord(i)-1) in str1:
                c+=1
            if i=='a' and 'z' in str1:
                c+=1
        return c==len(l)

# 48ms 17.5MB
class Solution:
    def canMakeSubsequence(self, source: str, target: str) -> bool:
        targetIdx, targetLen = 0, len(target)  
        for currChar in source:
            if targetIdx < targetLen and (ord(target[targetIdx]) - ord(currChar)) % 26 < 2:
                targetIdx += 1  
        return targetIdx == targetLen

File: JavaScript/WebsiteCode/LeechCode/Medium/MakeStringASubsequenceUsingCyclicIncrements.js
// JavaScript

/**
 * @param {string} str1
 * @param {string} str2
 * @return {boolean}
 */
// 8ms 55.62MB
var canMakeSubsequence = function(str1, str2) {
    let p1 = 0;
    let p2 = 0;

    while (p1 !== str1.length && p2 !== str2.length) {
        if (str1[p1] === str2[p2] || str1[p1] === Rotate(str2[p2])) {
            p1++;
            p2++;
        } else {
            p1++;
        }
    }

    return p2 === str2.length;
};

var Rotate = function(c) {
    if (c === 'a') {
        return 'z';
    }

    return String.fromCharCode(c.charCodeAt(0) - 1);
}

// 17ms 56.11MB
var canMakeSubsequence = function(source, target) {
    let targetIdx = 0;
    const targetLen = target.length;
    
    for (let currChar of source) {
        if (targetIdx < targetLen && 
            ((target.charCodeAt(targetIdx) - currChar.charCodeAt(0) + 26) % 26 <= 1)) {
            targetIdx++;
        }
    }
    
    return targetIdx === targetLen;
};